inventory.py

```diff
@@ -69,6 +76,77 @@ def slim_down_course_data(course_data: Sequence[Dict]) -> Sequence[Dict]:
         slim_course_dicts.append(slim_course_dict)
     return slim_course_dicts
 
+def get_next_page_url(response):
+    """
+    get the next page url from the Http response headers
+    :param response:
+    :type response: requests
+    :return: next_page_url
+    :rtype: str
+    """
+    logging.debug(get_next_page_url.__name__ + '() called')
+
+    if not response.links:
+        logging.debug('The api call do not have Link headers')
+        return None
+
+    for page in response.links:
+        if 'next' in page:
+            return response.links['next']['url']
+
+
+def handle_request_if_failed(response):
+    if response.status_code != 200:
+        return False
+    else:
+        return True
+
+
+def get_course_publish_date(df_row: str, publish_date, next_page_url = None)->str:
+    logger.debug(df_row)
+    if df_row['course_workflow_state'] != 'available':
+        return 'null'
+
+    course_id= df_row['course_id']
+
+    if next_page_url is not None:
+        url = next_page_url
+    else:
+        url = f"{API_SCOPE_PREFIX}/audit/course/courses/{course_id}?per_page={PER_PAGE_COUNT}"
+
+    try:
+        response = make_audit_logs_api_request(url)
+
+    except (Exception) as e:
+        logger.exception('getting published date for a course has erroneous response ' + e.message)
+        return publish_date
+
+    if not handle_request_if_failed(response):
+        return publish_date
+
+    audit_events = json.loads(response.text.encode('utf8'))
+    if not audit_events:
+        return publish_date
+
+    events = audit_events['events']
+
+    for event in events:
+        if event['event_type'] == 'published':
+            publish_date = event['created_at']
+            logger.info(f"Date for Workflow type :{df_row['course_workflow_state']} for course {course_id}")
+            break
+    if publish_date is None:
```

Comment:
so the logic is keep looking for the "published" event, until we find it; otherwise, we can stop there.

I understand the people can iterator through publish vs unpublish status, so this approach will get the earliest publish date. That is fine with me.

In this case, there is then no need to have the "publish_datet" param inside the get_course_publish_date() function, since the recursive function will stop once the date is found.

Reply:
see my comment here https://github.com/tl-its-umich-edu/course-inventory/pull/9#issuecomment-600088985